### PR TITLE
Add xpu2 compiler

### PR DESCRIPTION
**Pull request by Liu-xiandong**
### PR types

New features
### PR changes

Others
### Describe

This PR adds the source code compilation mode of XPU2, and the compilation commands are as follows:
```
cmake .. -DPY_VERSION=3.7 -DWITH_XPU_KP=ON -DWITH_GPU=OFF -DWITH_TESTING=ON -DCMAKE_BUILD_TYPE=Release
```

Its realization mainly involves the following aspects:

**1. Support source code compilation mode of XPU2 code:**

Mainly related to the modification of cmake, the main files involved are: cmake/generic.cmake cmake/operators.cmake cmake/xpu_kp.cmake
The specific calling process is as follows:
image

**2. Compatible operator mapping mode and source code compilation mode**

After adding the source code compilation mode, there are two XPU kernels for the same OP.
In order to be compatible with the two modes, three aspects of work have been done.

- Added a new library_type to distinguish the original mode. 
The main files involved are: _paddle/fluid/framework/library_type.h_

- In order to dynamically select the kernel at runtime, the flag and the list involved in the KP mode are added.
The main files involved are: _paddle/fluid/platform/flags.cc paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h paddle/fluid/platform/device/xpu/xpu_op_list.cc_

- Increase specific runtime control methods
The main files involved are: 
_paddle/fluid/framework/operator.cc_ (static graph) 
_paddle/fluid/imperative/prepared_operator.cc_ (dynamic graph)

The control graph of runtime is as follows:
image

### CODE REVIEW

File: cmake/xpu_kp.cmake [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
...ern_boost -I${CMAKE_SOURCE_DIR}/build/third_party/eigen3/src/extern_eigen3 -I${CMAKE_SOURCE_DIR}/build/third_party/threadpool/src/extern_threadpool -I${CMAKE_SOURCE_DIR}/build/third_party/dlpack/src/extern_dlpack/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/xxhash/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/warpctc/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/utf8proc/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/openblas/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/protobuf/include -I/usr/include/python3.7m -I/usr/local/lib/python3.7/dist-packages/numpy/core/include -I${CMAKE_SOURCE_DIR}/build/third_party/pybind/src/extern_pybind/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/gtest/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/xpu/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/gloo/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/xbyak/include -I${CMAKE_SOURCE_DIR}/build/third_party/install/xbyak/include/xbyak -I${CMAKE_SOURCE_DIR}/build/third_party/install/cryptopp/include -I${CMAKE_SOURCE_DIR}/build/third_party/pocketfft/src -I${CMAKE_SOURCE_DIR} -I${CMAKE_SOURCE_DIR}/paddle/fluid/platform)
```
Comment (xingfeng01) [inline cmake/xpu_kp.cmake]: 这里需要简化

File: cmake/xpu_kp.cmake [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  set(XPU_CXX_FLAGS  -Wno-error=pessimizing-move -Wno-error=constant-conversion -Wno-error=c++11-narrowing -Wno-error=shift-count-overflow -Wno-error=unused-local-typedef -Wno-error=deprecated-declarations -Wno-deprecated-declarations -std=c++14 -m64 -fPIC -fno-omit-frame-pointer  -Wall -Wno-inconsistent-missing-override -Wextra -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wno-unused-parameter -Wno-unused-function  -Wno-error=unused-local-typedefs -Wno-error=ignored-attributes  -Wno-error=int-in-bool-context -Wno-error=parentheses -Wno-error=address -Wno-ignored-qualifiers -Wno-ignored-attributes -Wno-parentheses -O3 -DNDEBUG )
```
Comment (xingfeng01) [inline cmake/xpu_kp.cmake]: 需要确认 flags

File: paddle/fluid/platform/aligned_vector.h [hunk 3]
```diff
@@ -15,7 +15,7 @@ limitations under the License. */
 #pragma once
 
 #include "paddle/fluid/platform/hostdevice.h"
-
+#define CHAR_BIT 8
```
Comment (xingfeng01) [inline paddle/fluid/platform/aligned_vector.h]: 需要 xpu 宏保护吗？

File: paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+namespace platform {
+
+using vartype = paddle::framework::proto::VarType;
+using pOpKernelType = paddle::framework::OpKernelType;
+using XPUKernelSet =
+    std::unordered_set<pOpKernelType, paddle::framework::OpKernelType::Hash>;
+using XPUOpMap = std::unordered_map<std::string, XPUKernelSet>;
+
+XPUOpMap& get_kp_ops() {
+  // KL1支持的op，通过op_name, data_type, place来索引
```
Comment (xingfeng01) [inline paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h]: 去掉中文注释

File: paddle/fluid/platform/device/xpu/xpu_op_list.cc [hunk 5]
```diff
@@ -75,6 +76,49 @@ bool is_in_xpu_black_list(const std::string& op_name) {
...
+  static std::mutex s_mtx;
+  if (!inited) {
+    std::lock_guard<std::mutex> guard(s_mtx);
+    if (!inited) {
+      if (std::getenv("XPU_KPWHITE_LIST") != nullptr) {
+        std::string ops(std::getenv("XPU_KPWHITE_LIST"));
+        tokenize(ops, ',', &xpu_kpwhite_list);
+      }
+      inited = true;
+      VLOG(3) << "XPU kpwhite List: ";
```
Comment (xingfeng01) [inline paddle/fluid/platform/device/xpu/xpu_op_list.cc:106]: 只是用来调试是吗

File: paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  static XPUOpMap s_xpu_kp_kernels{
+      {"elementwise_add_grad",
+       XPUKernelSet({pOpKernelType(vartype::FP32, XPUPlace()),
+                     pOpKernelType(vartype::FP16, XPUPlace())})},
+      {"elementwise_add",
+       XPUKernelSet({pOpKernelType(vartype::FP32, XPUPlace()),
+                     pOpKernelType(vartype::FP16, XPUPlace())})},
+      // AddMore
```
Comment (xingfeng01) [inline paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h]: 这个需要删除是吗 ？

File: cmake/xpu_kp.cmake [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    )
+    list(APPEND xpu_kernel_depends kernel_build/${kernel_name}.bin.o)
+
+  add_custom_command(
+    OUTPUT
+      kernel_build/${kernel_name}.host.o
+    COMMAND
+      ${CMAKE_COMMAND} -E make_directory kernel_build
+    COMMAND
+    ${XPU_CLANG} --sysroot=${CXX_DIR}  -O2 -fno-builtin -g -mcpu=xpu2  -fPIC ${XPU_CXX_DEFINES}  ${XPU_CXX_FLAGS} ${XPU_CXX_INCLUDES}
```
Comment (xingfeng01) [inline cmake/xpu_kp.cmake]: 注意 flags 比如 O2 -g 等

File: paddle/fluid/platform/aligned_vector.h (hunk 3 above)
Comment (Liu-xiandong) [reply paddle/fluid/platform/aligned_vector.h]: 这个不需要了

File: paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h (hunk 4 above)
Comment (Liu-xiandong) [reply paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h]: done

File: paddle/fluid/platform/device/xpu/xpu_op_list.cc (hunk 5 above)
Comment (Liu-xiandong) [reply paddle/fluid/platform/device/xpu/xpu_op_list.cc:106]: Yes

File: paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h (hunk 6 above)
Comment (Liu-xiandong) [reply paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h]: Done

File: cmake/xpu_kp.cmake (hunk 7 above)
Comment (Liu-xiandong) [reply cmake/xpu_kp.cmake]: OK

File: cmake/xpu_kp.cmake (hunk 1 above)
Comment (Liu-xiandong) [reply cmake/xpu_kp.cmake]: Done.

File: cmake/xpu_kp.cmake (hunk 2 above)
Comment (Liu-xiandong) [reply cmake/xpu_kp.cmake]: Done

File: paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+using vartype = paddle::framework::proto::VarType;
+using pOpKernelType = paddle::framework::OpKernelType;
+using XPUKernelSet =
+    std::unordered_set<pOpKernelType, paddle::framework::OpKernelType::Hash>;
+using XPUOpMap = std::unordered_map<std::string, XPUKernelSet>;
+
+XPUOpMap& get_kp_ops() {
+  static XPUOpMap s_xpu_kp_kernels{
+      // TODO(Liu-xiandong)
```
Comment (xingfeng01) [inline paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h]: 删除这个注释

Comment (Liu-xiandong) [reply paddle/fluid/platform/device/xpu/xpu_op_kpfirst_list.h]: Done

**Review by xingfeng01 — DISMISSED** [review]

**Review by lanxianghit — DISMISSED** [review]

File: CMakeLists.txt [hunk 9]
```diff
@@ -43,6 +43,7 @@ option(WITH_ONEMKL      "Compile PaddlePaddle with oneMKL"              OFF)
 option(WITH_GPU         "Compile PaddlePaddle with NVIDIA GPU"          ${CUDA_FOUND})
 option(WITH_TENSORRT    "Compile PaddlePaddle with NVIDIA TensorRT"     OFF)
 option(WITH_XPU         "Compile PaddlePaddle with BAIDU KUNLUN XPU"    OFF)
+option(WITH_XPU_KP         "Compile PaddlePaddle with BAIDU XPU compiler "    OFF)
```
Comment (wanghuancoder) [inline CMakeLists.txt]: option(WITH_XPU_KP "Compile PaddlePaddle with BAIDU XPU compiler" OFF)

File: CMakeLists.txt [hunk 10]
```diff
@@ -59,6 +60,9 @@ include(generic)            # simplify cmake module
 if (WITH_GPU  AND WITH_XPU)
```
Comment (wanghuancoder) [inline CMakeLists.txt:60]: if (WITH_GPU AND WITH_XPU)

File: CMakeLists.txt (hunk 9 above)
Comment (Liu-xiandong) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 10 above)
Comment (Liu-xiandong) [reply CMakeLists.txt:60]: Done

**Review by chenwhql — APPROVED** [review]
LGTM for operator.h

**Review by wanghuancoder — APPROVED** [review]

**Review by xingfeng01 — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

**Pull request merged**